### ISSUE TITLE
mgl-pax fails to load - can't find .git/refs/heads/master

### ISSUE THREAD

**Issue by metayan** [melisgl/mgl-pax#11]
```lisp
(ql:quickload :mgl-pax)
```
gives the following:
```
The file #P"/Users/Shared/cl/scm/mgl-pax/.git/refs/heads/master"
does not exist:
  No such file or directory
   [Condition of type FILE-DOES-NOT-EXIST]

Backtrace:
  0: (SB-IMPL::FILE-PERROR #P"/Users/Shared/cl/scm/mgl-pax/.git/refs/heads/master" 2 FILE-DOES-NOT-EXIST)
  1: (SB-IMPL::%OPEN-ERROR #P"/Users/Shared/cl/scm/mgl-pax/.git/refs/heads/master" 2 :IGNORE-THIS-ARG :ERROR)
  2: (OPEN #P"/Users/Shared/cl/scm/mgl-pax/.git/refs/heads/master" :DIRECTION :INPUT
       :ELEMENT-TYPE BASE-CHAR :IF-EXISTS NIL :IF-DOES-NOT-EXIST NIL :EXTERNAL-FORMAT :DEFAULT
       :CLASS SB-SYS:FD-STREAM)
  3: (MGL-PAX::READ-FIRST-LINE #P"/Users/Shared/cl/scm/mgl-pax/.git/refs/heads/master")
  4: (MGL-PAX::GIT-VERSION #P"/Users/Shared/cl/scm/mgl-pax/.git/")
  5: (MGL-PAX:MAKE-GITHUB-SOURCE-URI-FN :MGL-PAX "https://github.com/melisgl/mgl-pax" :GIT-VERSION NIL)
  6: (MGL-PAX::PAX-PAGES)
```
`mgl-pax` is cloned in a directory which is linked from local-projects.
Its `.git/config` contains
```
[core]
        repositoryformatversion = 0
        filemode = true
        bare = false
        logallrefupdates = true
        ignorecase = true
        precomposeunicode = true
[remote "origin"]
        url = https://github.com/melisgl/mgl-pax
        fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
        remote = origin
        merge = refs/heads/master
```
`git status` in the mgl-pax directory gives
```
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```
`git log` starts with
```
commit 4ada6eb26364e71addb169ce58e4ba83bc7a8eaa (HEAD -> master, origin/master, origin/HEAD)
Author: Gábor Melis <mega@retes.hu>
Date:   Tue Feb 16 14:42:35 2021 +0000
```
Using SBCL 2.1.2 on macOS 10.15.7. Also tried with older versions of SBCL.
git version 2.30.2

**Comment by melisgl** [comment]
That's weird. Can you confirm that `cat .git/HEAD` prints `ref: refs/heads/master`, but `.git/refs/heads/master` does not exist?

**Comment by metayan** [comment]
Confirming.
`cat .git/HEAD`
```
ref: refs/heads/master
```
No `.git/refs/heads/master`.

It seems that one cannot take for granted that `.git/refs/heads/` contains a file named as the current branch.

A quick check with the repos that I have locally shows
`ls -1d */.git/refs/heads |wc -l`
    5691
`ls -1d */.git/refs/heads/* |wc -l`
     833

However, `.git/info/refs` contains
```
4ada6eb26364e71addb169ce58e4ba83bc7a8eaa	refs/heads/master
4ada6eb26364e71addb169ce58e4ba83bc7a8eaa	refs/remotes/origin/HEAD
092444e777fbb935e290fc51abe90ecd0a72efe4	refs/remotes/origin/gh-pages
4ada6eb26364e71addb169ce58e4ba83bc7a8eaa	refs/remotes/origin/master
```
so probably that should be used instead.

Managed to circumvent the issue by commenting out
https://github.com/melisgl/mgl-pax/blob/4ada6eb26364e71addb169ce58e4ba83bc7a8eaa/src/doc.lisp#L454

**Comment by melisgl** [comment]
I think accessing `.git/refs/heads/master` directly was to avoid having to invoke git. I suppose `git rev-parse refs/heads/master` is the right way to determine the commit. Does it that work in your repo?

**Comment by metayan** [comment]
Yes, that works for the `mgl-pax` repository here.
However, I've been tinkering a bit and put together something that works with both cases without invoking `git`.
I can make a PR.

**Comment by melisgl** [comment]
I think invoking git may be the right way. I read it somewhere (can't find the link now) that packed repos may not have refs/heads/ populated or something like that.

**Comment by metayan** [comment]
You're right about packed repos - `git gc` removes the files in `.git/refs/heads/`.
Added PR #12 - it solves this issue for the packed repos, even though calling `git` directly would be even more general.
Maybe something like
```
git rev-parse --symbolic-full-name @
git rev-parse @
```
or
```
git branch --show-current
git show --no-patch --format=%H HEAD
```
I'm not familiar with how easy it is to call `git` from Lisp on some systems, though - I use Linux and macOS.

**Comment by metayan** [comment]
Fixed with `info/refs` in #12 and then with `git rev-parse` in https://github.com/melisgl/mgl-pax/commit/ff1a42c2f57dddd4a7c0783fac3f988632d8e257 .
